src/node/lattice.rs

```diff
@@ -0,0 +1,518 @@
+    /// This function moves the passed event into the lattice, and inserts the appropriate edges to
+    /// existing events in the graph based on the partial order defined in
+    /// [`OperatorEvent`](../operator_event/struct.OperatorEvent.html).
+    pub async fn add_event(&self, event: OperatorEvent) {
+        // Take locks over everything.
+        let forest: &mut StableGraph<Option<OperatorEvent>, ()> = &mut *self.forest.lock().await;
+        let roots: &mut HashSet<NodeIndex<u32>> = &mut *self.roots.lock().await;
+        let run_queue: &mut VecDeque<NodeIndex<u32>> = &mut *self.run_queue.lock().await;
+
+        // Iterate through all the roots, and figure out where to add a dependency edge.
+        let mut dfs = DfsPostOrder::empty(&*forest);
+        let mut dependencies: HashSet<NodeIndex<u32>> = HashSet::new();
+        let mut add_edges_to: Vec<NodeIndex<u32>> = Vec::new();
+        let mut add_edges_from: Vec<NodeIndex<u32>> = Vec::new();
+        let mut remove_roots: Vec<(usize, NodeIndex<u32>)> = Vec::new();
+
+        for root in &*roots {
+            // This function maintains the stack when iterating over the graph so we do not have to
+            // check the same node in the graph twice.
+            dfs.move_to(*root);
+
+            // Retrieve the next node in DFS order from the forest.
+            while let Some(nx) = dfs.next(&*forest) {
+                // Check if any of the children of the current node in DFS traversal already have a
+                // dependency on the node we are trying to add.
+                let mut dependent_child: bool = false;
+                for child in forest.neighbors(nx) {
+                    if dependencies.contains(&child) {
+                        dependencies.insert(nx);
+                        dependent_child = true;
+                        break;
+                    }
+                }
+
+                // If the event we are trying to add has already been added as a dependency of one
+                // of the children of the DFS node, then move to the next root of the forest.
+                if dependent_child {
+                    break;
+                }
+
+                if let Some(node) = forest.node_weight(nx).unwrap().as_ref() {
+                    // If the event we are trying to add is dependent on the DFS node, then
+                    // add an edge from the DFS node to the event to be added.
+                    if event < *node || event == *node {
+                        // Check if any of the children of the DFS node should be dependent on the
+                        // event we are trying to add.
+                        let mut remove_edges: Vec<NodeIndex<u32>> = Vec::new();
+                        for child in forest.neighbors(nx) {
+                            let child_node: &OperatorEvent =
+                                &forest.node_weight(child).unwrap().as_ref().unwrap();
+                            if *child_node < event {
```

Comment:
If you don't want to deref:
```suggestion
                            if child_node < &event {
```